src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs

```diff
@@ -678,301 +553,475 @@ public Task<AnalysisResult> GetAnalysisResultAsync(SemanticModel model, TextSpan
+
             try
             {
-                var taskToken = Interlocked.Increment(ref _currentToken);
+                cancellationToken.ThrowIfCancellationRequested();
+
+                analysisScope = GetPendingAnalysisScope(analysisScope, _analysisResultBuilder);
+                if (analysisScope == null)
+                    return;
+
+                (compilation, driver, driverAllocatedFromPool) = await getCompilationAndDriverAsync().ConfigureAwait(false);
+
+                // Driver must have been initialized.
+                Debug.Assert(driver.IsInitialized);
+                Debug.Assert(!driver.WhenInitializedTask.IsCanceled);
+
+                (var analyzerActionCounts, var hasAnyActionsRequiringCompilationEvents) = await getAnalyzerActionCountsAsync(
+                    originalAnalyzers, driver, compilation, cancellationToken).ConfigureAwait(false);
+                Func<DiagnosticAnalyzer, AnalyzerActionCounts> getAnalyzerActionCounts = analyzer => analyzerActionCounts[analyzer];
 
-                var pendingAnalyzers = _analysisResultBuilder.GetPendingAnalyzers(analysisScope.Analyzers);
-                if (pendingAnalyzers.Length > 0)
+                // We use different analyzer execution models based on whether we are analyzing the full compilation or not.
+                if (!analysisScope.IsSingleFileAnalysis)
+                {
+                    // We are performing full compilation analysis.
+                    // PERF: For improved performance, we first attach the event queue to the driver and then
+                    //       invoke compilation.GetDiagnostics() to ensure that we generate and process the
+                    //       compilation events concurrently.
+
+                    // Attach the driver to compilation and start processing events.
+                    // If we do not have any analyzer actions that are driven by compilation events,
+                    // we are only going to enqueue a CompilationStartedEvent, so ensure we
+                    // pass 'usingPrePopulatedEventQueue = true' for that case.
+                    driver.AttachQueueAndStartProcessingEvents(compilation.EventQueue!, analysisScope, usingPrePopulatedEventQueue: !hasAnyActionsRequiringCompilationEvents, cancellationToken);
+
+                    // If we have any analyzer actions that are driven by compilation events, we need to
+                    // force compilation diagnostics to populate the compilation event queue and force it to be completed.
+                    if (hasAnyActionsRequiringCompilationEvents)
+                        _ = compilation.GetDiagnostics(cancellationToken);
+
+                    // Wait for analyzer execution to complete.
+                    await driver.WhenCompletedTask.ConfigureAwait(false);
+
+                    _analysisResultBuilder.ApplySuppressionsAndStoreAnalysisResult(analysisScope, driver, compilation, getAnalyzerActionCounts);
```

Comment:
All of this is existing code (line 436 - line 452), just moved into this method. This allows us to unify the core analyzer computation logic for both full compilation diagnostic requests and single file/span diagnostic requests into this single method and avoid lot of duplicated code.